cli/command/environment_test.go

```diff
@@ -0,0 +1,381 @@
+	testCases := []struct {
+		name  string
+		flags map[string]interface{}
+		wait  bool
+	}{
+		{
+			name: "Wait",
+			wait: true,
+		},
+		{
+			name:  "NoWait",
+			flags: map[string]interface{}{"no-wait": true},
+			wait:  false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client, resolver, command, ctrl := initEnvCommandTest(t)
+			defer ctrl.Finish()
+
+			job := &models.Job{
+				JobID:  "job-id",
+				Status: job.Completed.String(),
+				Result: "entity-id",
+			}
+
+			resolver.EXPECT().
+				Resolve("environment", "name").
+				Return([]string{"id"}, nil)
+
+			client.EXPECT().
+				UpdateEnvironment(gomock.Any()).
+				Return(job.JobID, nil)
+
+			if tc.wait {
+				client.EXPECT().
+					ReadJob(job.JobID).
+					Return(job, nil)
+
+				client.EXPECT().
+					ReadEnvironment(job.Result).
+					Return(&models.Environment{}, nil)
+			}
+
+			c := GetCLIContext(t, []string{"name", "2"}, tc.flags)
+			if err := command.update(c); err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+func TestEnvironmentLink(t *testing.T) {
+	client, resolver, command, ctrl := initEnvCommandTest(t)
+	defer ctrl.Finish()
+
+	resolver.EXPECT().
+		Resolve("environment", "name1").
+		Return([]string{"id1"}, nil)
+
+	resolver.EXPECT().
+		Resolve("environment", "name2").
+		Return([]string{"id2"}, nil)
+
+	client.EXPECT().
+		CreateLink("id1", "id2").
+		Return(nil)
+
+	c := GetCLIContext(t, []string{"name1", "name2"}, nil)
+	if err := command.link(c); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentUnlink(t *testing.T) {
+	client, resolver, command, ctrl := initEnvCommandTest(t)
+	defer ctrl.Finish()
+
+	resolver.EXPECT().
+		Resolve("environment", "name1").
+		Return([]string{"id1"}, nil)
+
+	resolver.EXPECT().
+		Resolve("environment", "name2").
+		Return([]string{"id2"}, nil)
+
+	client.EXPECT().
+		DeleteLink("id1", "id2").
+		Return(nil)
+
+	c := GetCLIContext(t, []string{"name1", "name2"}, nil)
+	if err := command.unlink(c); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentLink_duplicateEnvironmentID(t *testing.T) {
+	_, resolver, command, ctrl := initEnvCommandTest(t)
+	defer ctrl.Finish()
+
+	resolver.EXPECT().
+		Resolve("environment", gomock.Any()).
+		Return([]string{"id1"}, nil).
+		Times(2)
+
+	c := GetCLIContext(t, []string{"name1", "name2"}, nil)
+	if err := command.link(c); err == nil {
+		t.Fatal("error was nil!")
+	}
+}
+
+func TestEnvironmentUnlink_duplicateEnvironmentID(t *testing.T) {
+	_, resolver, command, ctrl := initEnvCommandTest(t)
+	defer ctrl.Finish()
+
+	resolver.EXPECT().
+		Resolve("environment", gomock.Any()).
+		Return([]string{"id1"}, nil).
+		Times(2)
+
+	c := GetCLIContext(t, []string{"name1", "name2"}, nil)
+	if err := command.unlink(c); err == nil {
+		t.Fatal("error was nil!")
+	}
+}
+
+func initEnvCommandTest(t *testing.T) (*mock_client.MockClient, *mock_resolver.MockResolver, *EnvironmentCommand, *gomock.Controller) {
```

Comment:
I disagree - why repeat the same code in each test when you can encapsulate it in a single method call? The setup itself isn't relevant to the test, so minimizing the amount of setup lines seems better to me.

```
command, resolver, command, ctrl := initEnvCommandTest(t)
defer ctrl.Finish()

```
vs
```
base, ctrl := newTestCommand(t)
defer ctrl.Finish()
...
command := NewEnvironmentCommand(base.Command())

// each reference would then have to be prefixed by tc.
tc.Client
tc.Resolver
```

Reply:
because the point of the `BaseTestCommand` struct/pattern is encapsulation

Reply:
not sure what you mean; the `BaseTestCommand` does encapsulate the mock objects but why do we need to use an instance of BaseTestCommand rather than the properties of that object directly?

Reply:
> I don't see why returning multiple values and explicitly choosing to ignore as needed is an issue?

It's tedious to have to have to assign/ignore those variables as you work your way through writing a a unit test. 

That being said - I think consistency is more important than whatever pattern we choose. The difference it pretty small either way. I'm happy to adopt this pattern if the rest of the team is convinced it's the better option. 

To me, it seems like this is introducing a new pattern for the sake of introducing a new pattern. I don't view creating a helper function to save one line as much of a 'win',  neither do I see prefixing the mocks with `base.` as much of a 'loss'.

Reply:
cc @diemonster  @tlake @jparsons04 @nehayward

Reply:
My $0.02: I tend to weigh readability and explicitness far heavier than economy of lines. To that end, I think that declaring the commands in each test, while a little more verbose, is a fair bit more-readily understood. I'd vote ~80% in favor of not adding these helper functions.